File: problem-solving/baekjoon/17212_달나라_토끼를_위한_구매대금_지불_도우미.py
# https://www.acmicpc.net/problem/17212
import sys

dp = []
coins = [7, 5, 2, 1]

if __name__ == "__main__":
    n = int(sys.stdin.readline())
    dp = [100000 for _ in range(n+1)]
    dp[0] = 0
    for i in range(1, n+1):
        for coin in coins:
            if i-coin >= 0 and dp[i-coin]+1 < dp[i]:
                dp[i] = dp[i-coin] + 1
    print(dp[n])

File: problem-solving/baekjoon/17212_달나라_토끼를_위한_구매대금_지불_도우미.go
// https://www.acmicpc.net/problem/17212
package main

import (
	"bufio"
	"fmt"
	"os"
)

var (
	dp    []int
	coins = []int{7, 5, 2, 1}
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()

	var n int
	fmt.Fscanln(reader, &n)
	dp = make([]int, n+1)
	for i := 1; i < n+1; i++ {
		dp[i] = 100000
	}
	for i := 1; i < n+1; i++ {
		for _, coin := range coins {
			if i-coin >= 0 && dp[i-coin]+1 < dp[i] {
				dp[i] = dp[i-coin] + 1
			}
		}
	}
	fmt.Fprintln(writer, dp[n])
}
